Dockerfile

```diff
@@ -0,0 +1,40 @@
+# This Dockerfile builds an ruby environment for jekyll that empowers
+# making updates to the accumulo website without requiring the dev
+# to maintain a local ruby development environment.
+
+FROM ruby:2.7.8-slim-bullseye as base
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    build-essential \
+    git \
+    curl \
+    && rm -rf /var/lib/apt/lists/*
+
+WORKDIR /site
+
+
+# Copy over the Gemfiles so that all build dependencies are installed
+# during build vs at runtime.
+
+COPY Gemfile /site/Gemfile
+COPY Gemfile.lock /site/Gemfile.lock
```

Comment:
I'm a little confused by the name of this directory and the comment here.

First, the name of the directory is "site", which is a bit confusing, because later, the site is built in `_site`. Would that location be `/site/_site/` then?

Second, when is the rest of the website files copied over? Are those also copied over to the same `/site` directory inside the container? How does that work? Are they copied over to `/site` after the RUN commands but before the CMD commands?

And third, just to clarify, the RUN command is using the Gemfiles to install the gems when the container image is first created and started (using `docker run`) and then the CMD commands are run. For subsequent stopping (`docker stop`) and restarting (`docker start`) , only the CMD commands are executed on startup? Is this correct?

Reply:
> First, the name of the directory is "site", which is a bit confusing, because later, the site is built in `_site`. Would that location be `/site/_site/` then?

Yes! 

So three directories exist
1. $PWD (root of the repository on the host)
2. $PWD/_site (location of rendered changes on the host)
3. /site (working directory of the container)

These locations are mounted at container runtime to allow the webdev and webdev-validator containers to access the  source and rendered content. 

I completely agree that the `/site` directory name is a bit overused. So I'm happy to take suggestions on a better name for the container side of things. 

> Second, when is the rest of the website files copied over? Are those also copied over to the same `/site` directory inside the container? How does that work? Are they copied over to `/site` after the RUN commands but before the CMD commands?

So everything that happens in the Dockerfile is part of the container build stage. 
This is completely separate from actually running the container. 

The `RUN` commands specified in this file are just for local actions we want to trigger as part of building a container. 
CMD is the default command the container will execute when a `docker run` command is executed.

The rest of the site is mounted as a read/write volume at runtime. The contents of `accumulo-website/` are mounted at the container's `/site` directory location for the webdev container.  

Once the webdev container is running, and the source directory is mounted, then the CMD statement finally executes and the rendered content is created on the containers /site/_site directory, which maps to the host's $PWD/_site directory. 

This results in the following dir structure: 

webdev's container's filesystem
/site
  - contents of accumulo-website
  - /_site 
     - Rendered html from jekyll 

This ensures that no site files are embedded in a container and that the source repository remains the only location for changes to occur. 

Now, we want to mount the rendered content into the validator container. However, the webdev-validator container needs to have a slightly modified Gemfile. 

To accomplish this, the run command only mounts $PWD/_site to /site/_site. This sets up the following structure

webdev-validator's filesystem
/site
  - Gemfile (modified by container)
  - Gemfile.lock (modified by container)
  - /_site
     - Rendered html from jekyll

> And third, just to clarify, the RUN command is using the Gemfiles to install the gems when the container image is first created and started (using `docker run`) and then the CMD commands are run. For subsequent stopping (`docker stop`) and restarting (`docker start`) , only the CMD commands are executed on startup? Is this correct?

The RUN command is only triggered during a `docker build` command. 
The container is never started during the `docker build` stage. 

The CMD statement ONLY runs on a `docker run`  or `docker start` command and is never triggered during a `docker build` stage, only specified. 

If it helps to think about it this way, CMD and ENTRYPOINT are like environment variables I am setting in my Dockerfile. I'm specifying the future behavior of my container when it runs. So, I can define those variables at any point in my Dockerfile, their placement in the Dockerfile has no bearing on what actions happen during container build. 

Whereas RUN statements are actual, order dependent commands that are used to trigger immediate actions in my container during the build stage to modify the container in some way.

Reply:
> 2. $PWD/_site (location of rendered changes on the host)

Gotcha. Rendered in the container process/memory space, but written to disk on the host because that is mounted.

> I completely agree that the `/site` directory name is a bit overused. So I'm happy to take suggestions on a better name for the container side of things.

Maybe "/mnt/repo" or "/mnt/staging" or "/mnt/build" or "/mnt/workdir"? I like putting things in `/mnt` because it makes things pretty clear that it's mounted, and from the perspective from inside the container, this is a pretty standard location for external filesystems to be mounted at.

> So everything that happens in the Dockerfile is part of the container build stage. This is completely separate from actually running the container.

My question isn't really about WHEN the RUN commands execute (I get that it's during the build stage), but WHERE (because it doesn't make sense that they would run locally on the host). My understanding is that the build stage involves running a container from the base image, executing RUN commands to modify the image, and when that build container stops, the resulting image is the modified image that we use to launch our own containers.

Essentially:

```
docker build -> create temporary build container from base image, start build container, exec RUN commands inside build container, stop build container, save modified image, delete build container
docker create -> create container from image
docker start -> run container with specified command, or stored CMD if command is unspecified
docker run -> shorthand for docker create + docker start
```

> This ensures that no site files are embedded in a container and that the source repository remains the only location for changes to occur.

Some rendering might use /tmp inside the container. I'm not sure if any of that is preserved inside the container, but I don't think we care if it is or isn't.

> Now, we want to mount the rendered content into the validator container. However, the webdev-validator container needs to have a slightly modified Gemfile.
> 
> To accomplish this, the run command only mounts $PWD/_site to /site/_site. This sets up the following structure

My understanding here is that the build of the validator container does *not* have the repo mounted. So, the Gemfiles it is modifying are the ones copied to the base image when that was built.

The confusion here is that when we run a container from the base image, we're mounting over the same directory as we copied these files. Are these just masked/hidden by the mount, then? Usually it's a good idea to mount to an empty directory. But in this case, it seems we're not doing that in the base image.

If we're mounting over the copied Gemfiles, then there's no reason to create a second validator image... we can just modify the Gemfile during the base build, install *all* the gems to the image during the base image build. When we mount our website repo to render the Jekyll, we hide the modified Gemfiles. When we mount our rendered `_site` directory to do the validation, we re-expose the modified Gemfiles. We don't need a second image at all... we can do everything we want with the single image.

Actually, we don't even need to copy over the Gemfiles during the build stage... those aren't needed for the htmlproofer. htmlproofer has need of its own Gemfiles. We're only copying over the ones from the repo and modifying them... but we don't need to do that... we can just have a separate set for the htmlproofer... and we don't even need the lock one at all for that.

I think we can simplify things dramatically by just using one container.

I do have one more question about WORKDIR... it seems that's overloaded to specify the current working directory inside the build container as well as the runtime container. But, I don't see why that necessarily needs to be the case.

Also, one thing I was thinking was that it can be really hard to verify links using the rendered HTML, because you don't have an absolute directory to resolve absolute paths. The generated site is intended to be served at the root of a webserver. I don't know if htmlproofer supports this, but it'd probably be better to run the validation against `http://localhost:4000/` rather than `file:///site/_site/`.

If we want to get this in quickly, we could get the Dockerfile in without the validator stuff right away. I think that because the validator stuff adds so much complexity (extra steps, extra documentation, maybe extra containers) that still needs polishing, I want to go back to my much earlier opinion that the validation stuff should be added in a separate PR. At this point, I think it's likely to hold up this PR further.

Reply:
> Maybe "/mnt/repo" or "/mnt/staging" or "/mnt/build" or "/mnt/workdir"? I like putting things in `/mnt` because it makes things pretty clear that it's mounted, and from the perspective from inside the container, this is a pretty standard location for external filesystems to be mounted at.

Out of those options I like `/mnt/workdir` the best. 

> 
> > So everything that happens in the Dockerfile is part of the container build stage. This is completely separate from actually running the container.
> 
> My question isn't really about WHEN the RUN commands execute (I get that it's during the build stage), but WHERE (because it doesn't make sense that they would run locally on the host). My understanding is that the build stage involves running a container from the base image, executing RUN commands to modify the image, and when that build container stops, the resulting image is the modified image that we use to launch our own containers.
> 
> Essentially:
> 
> ```
> docker build -> create temporary build container from base image, start build container, exec RUN commands 
> inside build container, stop build container, save modified image, delete build container
> docker create -> create container from image
> docker start -> run container with specified command, or stored CMD if command is unspecified
> docker run -> shorthand for docker create + docker start
>```

RUN commands are always run in a docker image and are executed in the WORKDIR that has been specified in the Dockerfile. 
https://docs.docker.com/engine/reference/builder/#run 

For now, realize that you don't have a container "running" as part of the build process.
Instead you have something similar to a git tree where each command in a dockerfile is executed by the docker deamon and results in a new commit layer being added (ENV, LABEL, and WORKDIR are special and don't create a new commit layer).  That new layer is then used by the next command as it's container context.

https://docs.docker.com/build/building/packaging/#dockerfile 

The final image layer generated is the last "layer" of changes that comprise the image and is tagged as the final image.

So it's different from building a virtual machine image where you only get an single output of a image based on all the commands run within the image at build.

This allows Docker to share image layers between containers and save space. 

> Some rendering might use /tmp inside the container. I'm not sure if any of that is preserved inside the container, but I don't think we care if it is or isn't.

The github action only referenced publishing the contents of `_site` so I don't think the possible use of `/tmp` matters since it was never published before.
https://github.com/apache/accumulo-website/blob/0d3904b0c3b4eb839bb59158946d9068fba7ef53/.github/workflows/jekyll.yaml#L51-L56


> > Now, we want to mount the rendered content into the validator container. However, the webdev-validator container needs to have a slightly modified Gemfile.
> > To accomplish this, the run command only mounts $PWD/_site to /site/_site. This sets up the following structure
> 
> My understanding here is that the build of the validator container does _not_ have the repo mounted. So, the Gemfiles it is modifying are the ones copied to the base image when that was built.

Correct. 

> 
> The confusion here is that when we run a container from the base image, we're mounting over the same directory as we copied these files. Are these just masked/hidden by the mount, then? Usually it's a good idea to mount to an empty directory. But in this case, it seems we're not doing that in the base image.

I agree that it's a good idea to mount to an empty directory for traditional systems. 

However, mapping an existing directory overtop of a pre-defined one is a standard action with containers. 
Docker just discards the contents of the container directory and references the host dir.

> 
> If we're mounting over the copied Gemfiles, then there's no reason to create a second validator image... we can just modify the Gemfile during the base build, install _all_ the gems to the image during the base image build. When we mount our website repo to render the Jekyll, we hide the modified Gemfiles. When we mount our rendered `_site` directory to do the validation, we re-expose the modified Gemfiles. We don't need a second image at all... we can do everything we want with the single image.
> 
> Actually, we don't even need to copy over the Gemfiles during the build stage... those aren't needed for the htmlproofer. htmlproofer has need of its own Gemfiles. We're only copying over the ones from the repo and modifying them... but we don't need to do that... we can just have a separate set for the htmlproofer... and we don't even need the lock one at all for that.
> 
> I think we can simplify things dramatically by just using one container.

My original design had the single container but I had split it into two to restrict the Gemfile changes. 
So if there's a better way to separate Gemfile changes I'm happy to try that. 
I do agree that it massively simplifies things. 

> 
> I do have one more question about WORKDIR... it seems that's overloaded to specify the current working directory inside the build container as well as the runtime container. But, I don't see why that necessarily needs to be the case.

I'm not exactly sure what question you're asking here, but I'll try and provide the information I know. 
 
So WORKDIR sets the PWD for any following RUN, COPY, ENTRYPOINT, and CMD commands in the Dockerfile. 

If you don't set WORKDIR explicitly then it will default to `/` if it was not previously defined by the image in your FROM command.
 
Typically you would set the WORKDIR to ensure that all of your operations are running from an intended place, vs happening in unexpected locations within the container. 

If you wanted your runtime container to run things in a different location than the build commands, you can simply add another WORKDIR line in your Dockerfile and any commands defined after it will use the new location (like CMD). 

https://docs.docker.com/engine/reference/builder/#workdir 

> 
> Also, one thing I was thinking was that it can be really hard to verify links using the rendered HTML, because you don't have an absolute directory to resolve absolute paths. The generated site is intended to be served at the root of a webserver. I don't know if htmlproofer supports this, but it'd probably be better to run the validation against `http://localhost:4000/` rather than `file:///site/_site/`.
> 
htmlproofer seems explicitly designed to run against rendered HTML code in a directory. I didn't see anything in the documentation related to pointing it at a live site.
The project does support [adjusting for a baseurl](https://github.com/gjtorikian/html-proofer#adjusting-for-a-baseurl)

> If we want to get this in quickly, we could get the Dockerfile in without the validator stuff right away. I think that because the validator stuff adds so much complexity (extra steps, extra documentation, maybe extra containers) that still needs polishing, I want to go back to my much earlier opinion that the validation stuff should be added in a separate PR. At this point, I think it's likely to hold up this PR further.

At this point I agree. 

Figuring out how to support gem isolation is important but we can get the base image docker changes in and provide benefit for any developers generating documentation.

Reply:
> always run in a docker image

I think part of my earlier confusion comes from your loose interchange of "image" and "container". I think I know enough to parse those out without trouble now. But, for clarity, I don't think it makes sense to say anything is running in an image. I think if we were to be precise, then things run inside containers (as opposed to images), and containers are launched/created from an image. I think of images as just snapshots of a particular container's (stopped) state, from which another container can be launched. The build stage, therefore, just seems to be a starting up a container with one base image as input, and a new (possibly incremental CoW) image as output.

Reply:
> I don't know what benefit you get from being in a different directory that would justify the change.

Mostly, clarity in understanding which file was being edited (the copy or the mounted original). But I agree this doesn't matter much.